design/pipeline-artifacts.md

```diff
@@ -0,0 +1,145 @@
+# Pipeline Artifacts YAML shortcut
+
+**Status: PM spec**
+
+Pipeline Artifacts are the new way to move files between jobs and stages in your pipeline. They are hosted in Azure Artifacts and will eventually entirely replace FCS "Build Artifacts". Because moving files between jobs and stages is a crucial part of most CI/CD workflows, and because Pipeline Artifacts are expected to be the default way to do so, this spec proposes a YAML shortcut syntax for publishing and downloading artifacts.
+
+In this document, `artifact` and `downloadArtifact` refer specifically to Pipeline Artifacts. Artifacts are distinct from `resources`, which are various kinds of inputs like `pipelines`, `containers`, `repositories`, and `packages`. Ashok is driving the YAML spec for these inputs, and he and I (amullans) are in sync.
+
+## Proposal: artifact
+
+`artifact` is a shortcut for the [Publish Pipeline Artifacts](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/publish-pipeline-artifact.md) task. It will publish files from the current job to be used in subsequent jobs or in other stages or pipelines.
+
+### Schema
+
+```yaml
+- artifact: string | [ string ]  # a minimatch path or list of [minimatch paths](tasks/file-matching-patterns.md) to publish
+  name: string # identifier for this artifact (no spaces allowed), defaults to 'default'
+  prependPath: string # a directory path that will be prepended to all published files
+  seal: boolean # if true, finalizes the artifact so no more files can be added after this step
+```
+
+### Example
+
+```yaml
+- publishArtifact:
+  - **/bin/*
+  - **/obj/*
+  name: webapp
+  seal: true
+```
+
+---
+
+## Proposal: downloadArtifact
+
+`downloadArtifact` is a shortcut for the [Download Pipeline Artifacts](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/download-pipeline-artifact.md) task. It will download artifacts published from a previous job or stage or from another pipeline. Artifacts are downloaded either to `$PIPELINES_RESOURCESDIR` or to the directory specified in `root`.
+
+### Schema
+
+```yaml
+- downloadArtifacts:
```

Comment:
I want to flip this to match the `- artifact` pattern. Schema:
```yaml
- downloadArtifact: string # artifact name
  patterns: string | [ string ]
  ...
```
and you can say this multiple times as "steps".